### ISSUE TITLE
D3D_DestroyRenderer: Crash on attempt to destroy renderer

### ISSUE THREAD

**Issue by Wohlstand** [libsdl-org/SDL#5814]
Some days ago, one of my Windows users reported the crash of the game when attempting to quit it. Now, I attempted to run the thing, and I found the similar error was happen at me through the `SDL_DestroyRenderer()` call.

Error text:
```
HEAP[thextech.exe]: 
Heap block at 0000000008DF6730 modified at 0000000008DF6741 past requested size of 1
HEAP[thextech.exe]: 
Invalid address specified to RtlFreeHeap( 0000000008DF0000, 0000000008DF6740 )
```
- SDL2 version 2.0.22

On previous SDL2 error wasn't happen.

![изображение](https://user-images.githubusercontent.com/6751442/174335527-ca71eb2f-2df5-4ac8-80ff-b1fac8777382.png)

**Comment by slouken** [comment]
I'm not sure how you're getting into `Direct3DCreate9Ex()` from `D3D_DestroyRenderer()`, can you debug into it and see what's happening?

Also, does this happen with the latest code on GitHub?

**Comment by Wohlstand** [comment]
> I'm not sure how you're getting into Direct3DCreate9Ex() from D3D_DestroyRenderer(), can you debug into it and see what's happening?

The `IDirect3DDevice9_Release()` macro covers that :eyes:

> Also, does this happen with the latest code on GitHub?

I'll try after sleep.

**Comment by cgutman** [comment]
`IDirect3DDevice9_Release()` isn't actually calling `Direct3DCreate9Ex()`. The symbol resolver is just trying make sense of a stack frame with an address that is far away from any exported functions.

I believe it just matches the nearest preceding exported function to the address being resolved. It would be obvious that the function offset made no sense if you could see it, but VS chooses to hide that information for some reason.

**Comment by Wohlstand** [comment]
Okay, I ran the 2.23.1 version, and the crash is the same.

So, inside the `IDirect3DDevice9_Release` macro is the calling of the `(p)->Release()` thing (the pointer to an actual release call).

I see, it points the "Direct3DCreate9Ex" method here:
![изображение](https://user-images.githubusercontent.com/6751442/174503634-6c8f5a07-a420-4ee6-a2e6-4dcc9bd04a58.png)

that's why it got called.

Do you mean, there is a possible sort of memory damage happened that confused the pointer?

**Comment by cgutman** [comment]
No, the pointer is fine. The offset tells you that the symbol resolver is returning non-sense. A function pointer wouldn't point to a function with a non-zero offset (let alone an address 2KB from the start of the function).

The real function it's pointing to is not exported. Because you don't have symbols for d3d9.dll, you can't see what the real function is. The symbol resolver is just picking the nearest preceding exported function and computing the offset relative to that.

I recommend trying to [enable Page Heap](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/example-12---using-page-heap-verification-to-find-a-bug) on your process and see if that points you directly at the faulty code (which may indeed be VBoxDispD3D.dll).

**Comment by Wohlstand** [comment]
Some good news: at the side of my friend, the problem has been resolved after I updated the library to 2.23.1 the on-exit crash has gone. This problem looks like, really goes to another place, will try to debug somewhere deeper...

**Comment by slouken** [comment]
I'm glad the crash is resolved!
